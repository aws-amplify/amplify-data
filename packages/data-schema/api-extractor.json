{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/src/index.d.ts"
}
