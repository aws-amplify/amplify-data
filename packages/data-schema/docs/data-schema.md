<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
