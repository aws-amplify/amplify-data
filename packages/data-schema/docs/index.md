<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md)

## API Reference

## Packages

|  Package | Description |
|  --- | --- |
|  [@aws-amplify/data-schema](./data-schema.md) |  |

