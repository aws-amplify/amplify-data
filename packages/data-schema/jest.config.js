/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
  testPathIgnorePatterns: [
    ".test-d.ts", // don't directly execute these files. They'll be tested via their `test.ts` counterparts
    "/node_modules/",
    "utils.ts",
    "/fixtures/",
    "/MappedTypes/",
  ],
  coverageThreshold: {
		global: {
			branches: 72,
			functions: 86,
			lines: 85,
			statements: 85,
		},
	},
};
