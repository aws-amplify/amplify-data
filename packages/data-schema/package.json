{
  "name": "@aws-amplify/data-schema",
  "version": "0.14.31",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-api-next.git"
  },
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.mjs",
  "typings": "./dist/esm/index.d.ts",
  "scripts": {
    "build:esm-cjs": "rollup -c rollup.config.mjs",
    "build": "npm run clean && npm run build:esm-cjs",
    "clean": "rimraf dist tsconfig.tsbuildinfo",
    "docs": "api-extractor run && api-documenter markdown --input-folder temp --output-folder docs",
    "docs:temp": "api-extractor run && api-documenter markdown --input-folder temp --output-folder temp/docs",
    "check-api": "rimraf temp && npm run docs:temp && node ../../scripts/shallow-dir-diff.js docs temp/docs",
    "lint": "eslint .",
    "test": "jest"
  },
  "exports": {
    ".": {
      "types": "./dist/esm/src/index.d.ts",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.js",
      "node": "./src/index.ts"
    },
    "./internals": {
      "types": "./dist/esm/src/internals/index.d.ts",
      "import": "./dist/esm/internals/index.mjs",
      "require": "./dist/cjs/internals/index.js",
      "node": "./src/internals/index.ts"
    },
    "./runtime": {
      "types": "./dist/esm/src/runtime/index.d.ts",
      "import": "./dist/esm/runtime/index.mjs",
      "require": "./dist/cjs/runtime/index.js",
      "node": "./src/runtime/index.ts"
    },
    "./package.json": "./package.json"
  },
  "typesVersions": {
    ">=5.0": {
      "internals": [
        "./dist/esm/src/internals/index.d.ts"
      ]
    }
  },
  "dependencies": {
    "@aws-amplify/data-schema-types": "*",
    "@types/aws-lambda": "^8.10.134",
    "rxjs": "^7.8.1"
  },
  "devDependencies": {
    "@changesets/cli": "^2.26.2",
    "@rollup/plugin-typescript": "11.1.5",
    "@tsd/typescript": "^5.1.6",
    "@types/jest": "29.5.4",
    "jest": "^29.7.0",
    "jest-tsd": "^0.2.2",
    "rimraf": "^5.0.5",
    "rollup": "^4.9.6",
    "ts-jest": "^29.1.1",
    "turbo": "^1.10.14",
    "typescript": "^5.1.6"
  },
  "files": [
    "dist/**",
    "internals",
    "runtime",
    "package.json",
    "README.md",
    "CHANGELOG",
    "LICENSE",
    "NOTICE"
  ]
}
