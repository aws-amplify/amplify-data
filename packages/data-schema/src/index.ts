import { schema } from './ModelSchema';
import { model } from './ModelType';
import {
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
} from './ModelField';
import { ref } from './RefType';
import { hasOne, hasMany, belongsTo, manyToMany } from './ModelRelationalField';
import { allow } from './Authorization';
import { customType } from './CustomType';
import { enumType } from './EnumType';
import { query, mutation, subscription } from './CustomOperation';
import { handler } from './Handler';

// We are re-exporting with this pattern to make the JSDoc happy.
// Otherwise the JSDoc wouldn't consistently show up as a customer types
// "a."
export {
  schema,
  model,
  ref,
  customType,
  enumType as enum,
  query,
  mutation,
  subscription,
  hasOne,
  hasMany,
  belongsTo,
  manyToMany,
  allow,
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
  handler,
};
