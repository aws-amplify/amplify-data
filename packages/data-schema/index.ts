import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
