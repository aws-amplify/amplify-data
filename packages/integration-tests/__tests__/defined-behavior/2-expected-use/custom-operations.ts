import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import type {
  SelectionSet,
  Expect,
  Equal,
} from '@aws-amplify/data-schema-types';
import {
  buildAmplifyConfig,
  mockedGenerateClient,
  optionsAndHeaders,
  useState,
  expectGraphqlMatches,
} from '../../utils';

describe('custom operations', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const dummyHandler = '' as any;

  // #region covers d9c0f0f657dbbe8a, e394d9c98b1a1f8d
  const schema = a.schema({
    PhoneNumber: a
      .model({
        phoneNumber: a.string().required(),
      })
      .identifier(['phoneNumber'])
      .authorization((allow) => [allow.publicApiKey()]),
    checkBatchOfPhoneNumbersForActiveUsers: a
      .query()
      .arguments({
        phoneNumbers: a.string().array(),
      })
      .returns(a.ref('PhoneNumber').array())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [
        allow.groups(['admin', 'projectManager']),
        allow.publicApiKey(),
      ]),
    findJustOnePhoneNumber: a
      .query()
      .arguments({
        query: a.string(),
      })
      .returns(a.ref('PhoneNumber'))
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    echo: a
      .query()
      .arguments({
        value: a.string(),
      })
      .returns(a.string())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    EchoResult: a.customType({
      result: a.string(),
    }),
    echoWithCustomType: a
      .query()
      .arguments({
        value: a.string(),
      })
      .returns(a.ref('EchoResult'))
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    multiEcho: a
      .query()
      .arguments({
        value: a.string(),
      })
      .returns(a.ref('EchoResult').array())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    soloAsync: a
      .query()
      .arguments({
        value: a.string(),
      })
      .handler(a.handler.function(dummyHandler).async())
      .authorization((allow) => [allow.publicApiKey()]),
    asyncSync: a
      .query()
      .arguments({
        value: a.string(),
      })
      .handler([
        a.handler.function(dummyHandler).async(),
        a.handler.function(dummyHandler),
      ])
      .returns(a.ref('EchoResult'))
      .authorization((allow) => [allow.publicApiKey()]),
    syncAsync: a
      .query()
      .arguments({
        value: a.string(),
      })
      .handler([
        a.handler.function(dummyHandler),
        a.handler.function(dummyHandler).async(),
      ])
      .authorization((allow) => [allow.publicApiKey()]),
    syncSync: a
      .query()
      .arguments({
        value: a.string(),
      })
      .handler([
        a.handler.function(dummyHandler),
        a.handler.function(dummyHandler),
      ])
      .returns(a.ref('EchoResult'))
      .authorization((allow) => [allow.publicApiKey()]),
    asyncAsync: a
      .query()
      .arguments({
        value: a.string(),
      })
      .handler([
        a.handler.function(dummyHandler).async(),
        a.handler.function(dummyHandler).async(),
      ])
      .authorization((allow) => [allow.publicApiKey()]),
    CustomArgType: a.customType({
      message: a.string(),
      count: a.integer(),
    }),
    NestedObjectType: a.customType({
      innerField1: a.boolean(),
      innerField2: a.string(),
    }),

    NestedFieldType: a.customType({
      nestedObject1: a.ref('NestedObjectType'),
    }),
    queryWithCustomTypeArg: a
      .query()
      .arguments({
        customArg: a.ref('CustomArgType'),
      })
      .returns(a.string())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    mutateWithCustomTypeArg: a
      .mutation()
      .arguments({
        customArg: a.ref('CustomArgType'),
      })
      .returns(a.string())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    mutationWithNestedCustomType: a
      .mutation()
      .arguments({
        nestedField: a.ref('NestedFieldType'),
      })
      .returns(a.string())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    queryWithRefArg: a
      .query()
      .arguments({
        refArg: a.ref('EchoResult'),
      })
      .returns(a.string())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    mutationWithRefArg: a
      .mutation()
      .arguments({
        refArg: a.ref('EchoResult'),
      })
      .returns(a.string())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    ComplexCustomArgType: a.customType({
      field1: a.string(),
      field2: a.integer(),
    }),
    complexQueryOperation: a
      .query()
      .arguments({
        scalarArg: a.string(),
        customArg: a.ref('ComplexCustomArgType'),
        refArg: a.ref('EchoResult'),
      })
      .returns(a.string())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
    complexMutation: a
      .mutation()
      .arguments({
        scalarArg: a.string(),
        customArg: a.ref('ComplexCustomArgType'),
        refArg: a.ref('EchoResult'),
      })
      .returns(a.string())
      .handler(a.handler.function(dummyHandler))
      .authorization((allow) => [allow.publicApiKey()]),
  });

  type Schema = ClientSchema<typeof schema>;

  type ExpectedQueryWithCustomTypeArg = {
    customArg?: {
      message?: string | null;
      count?: number | null;
    } | null;
  };
  type ActualQuertWithCustomTypeArg = Schema['queryWithCustomTypeArg']['args'];
  type TestEchoWithCustomTypeArg = Expect<
    Equal<ActualQuertWithCustomTypeArg, ExpectedQueryWithCustomTypeArg>
  >;

  type ExpectedMutateWithCustomTypeArg = {
    customArg?: {
      message?: string | null;
      count?: number | null;
    } | null;
  };
  type ActualMutateWithCustomTypeArg =
    Schema['mutateWithCustomTypeArg']['args'];
  type TestMutateWithCustomTypeArg = Expect<
    Equal<ActualMutateWithCustomTypeArg, ExpectedMutateWithCustomTypeArg>
  >;

  type ExpectedNestedCustomTypeArgs = {
    nestedField?: {
      nestedObject1?: {
        innerField1?: boolean | null;
        innerField2?: string | null;
      } | null;
    } | null;
  };
  type ActualNestedCustomTypeArgs =
    Schema['mutationWithNestedCustomType']['args'];
  type TestNestedCustomTypeArgs = Expect<
    Equal<ActualNestedCustomTypeArgs, ExpectedNestedCustomTypeArgs>
  >;

  type ExpectedQueryWithRefArg = {
    refArg?: {
      result?: string | null;
    } | null;
  };
  type ActualQueryWithRefArg = Schema['queryWithRefArg']['args'];
  type TestQueryWithRefArg = Expect<
    Equal<ActualQueryWithRefArg, ExpectedQueryWithRefArg>
  >;

  type ExpectedMutationWithRefArg = {
    refArg?: {
      result?: string | null;
    } | null;
  };
  type ActualMutationWithRefArg = Schema['mutationWithRefArg']['args'];
  type TestMutationWithRefArg = Expect<
    Equal<ActualMutationWithRefArg, ExpectedMutationWithRefArg>
  >;

  type ExpectedComplexQueryArgs = {
    scalarArg?: string | null;
    customArg?: {
      field1?: string | null;
      field2?: number | null;
    } | null;
    refArg?: {
      result?: string | null;
    } | null;
  };
  type ActualComplexArgs = Schema['complexQueryOperation']['args'];
  type TestComplexArgs = Expect<
    Equal<ActualComplexArgs, ExpectedComplexQueryArgs>
  >;

  type ExpectedComplexMutationArgs = {
    scalarArg?: string | null;
    customArg?: {
      field1?: string | null;
      field2?: number | null;
    } | null;
    refArg?: {
      result?: string | null;
    } | null;
  };
  type ActualComplexMutationArgs = Schema['complexMutation']['args'];
  type TestComplexMutationArgs = Expect<
    Equal<ActualComplexMutationArgs, ExpectedComplexMutationArgs>
  >;
  // #endregion

  test('schema.transform() includes custom types, ref types, and operations', () => {
    const transformedSchema = schema.transform();
    const expectedTypes = ['CustomArgType', 'EchoResult', 'Query', 'Mutation'];
    const expectedOperations = [
      'queryWithCustomTypeArg(customArg: QueryWithCustomTypeArgCustomArgInput): String',
      'queryWithRefArg(refArg: QueryWithRefArgRefArgInput): String',
      'mutateWithCustomTypeArg(customArg: MutateWithCustomTypeArgCustomArgInput): String',
      'mutationWithRefArg(refArg: MutationWithRefArgRefArgInput): String',
    ];
    const expectedInputTypes = [
      'input QueryWithCustomTypeArgCustomArgInput',
      'input QueryWithRefArgRefArgInput',
      'input MutateWithCustomTypeArgCustomArgInput',
      'input MutationWithRefArgRefArgInput',
    ];

    expectedTypes.forEach((type) => {
      expect(transformedSchema.schema).toContain(`type ${type}`);
    });

    expectedOperations.forEach((operation) => {
      expect(transformedSchema.schema).toContain(operation);
    });

    expectedInputTypes.forEach((inputType) => {
      expect(transformedSchema.schema).toContain(inputType);
    });
  });
  test('primitive type result', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          echo: 'Echo result',
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);
    const client = generateClient<Schema>();
    // #region covers ffefd700b1e323c9
    const { data } = await client.queries.echo({ value: 'something' });
    // #endregion

    expect(data).toEqual('Echo result');
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('custom type result', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          echoWithCustomType: {
            result: 'custom type echo result',
          },
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);
    const client = generateClient<Schema>();

    const { data } = await client.queries.echoWithCustomType({
      value: 'something',
    });

    expect(data).toEqual({
      result: 'custom type echo result',
    });
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('custom type array result', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          multiEcho: [
            {
              result: 'CT echo result: hello',
            },
            null, // custom resolvers may return nulls, especially of piping data from dynamo
            {
              result: 'CT echo result: hello',
            },
            {
              result: 'CT echo result: hello',
            },
          ],
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);
    const client = generateClient<Schema>();

    const { data } = await client.queries.multiEcho({ value: 'something' });
    expect(data).toEqual([
      {
        result: 'CT echo result: hello',
      },
      {
        result: 'CT echo result: hello',
      },
      {
        result: 'CT echo result: hello',
      },
    ]);
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('model result', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          findJustOnePhoneNumber: {
            phoneNumber: '123-123-1234',
            userId: 'some-user-id',
            createdAt: '2024-05-17T22:16:17.712Z',
            updatedAt: '2024-05-17T22:16:17.712Z',
          },
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);
    const client = generateClient<Schema>();

    const { data } = await client.queries.findJustOnePhoneNumber({
      query: 'something',
    });
    expect(data).toEqual(
      expect.objectContaining({
        phoneNumber: '123-123-1234',
        userId: 'some-user-id',
        createdAt: '2024-05-17T22:16:17.712Z',
        updatedAt: '2024-05-17T22:16:17.712Z',
      }),
    );
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('model array result', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          checkBatchOfPhoneNumbersForActiveUsers: [
            null, // custom resolvers may return nulls, especially of piping data from dynamo
            {
              phoneNumber: '222-2222',
              userId: 'some-user-id',
              createdAt: '2024-05-17T22:15:58.521Z',
              updatedAt: '2024-05-17T22:15:58.521Z',
            },
            {
              phoneNumber: '333-3333',
              userId: 'some-other-user-id',
              createdAt: '2024-05-17T22:16:17.712Z',
              updatedAt: '2024-05-17T22:16:17.712Z',
            },
          ],
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);
    const client = generateClient<Schema>();

    const { data } =
      await client.queries.checkBatchOfPhoneNumbersForActiveUsers({
        phoneNumbers: ['123', '123'],
      });

    expect(data?.length).toEqual(2);
    expect(data?.[0]).toEqual(
      expect.objectContaining({
        phoneNumber: '222-2222',
        userId: 'some-user-id',
        createdAt: '2024-05-17T22:15:58.521Z',
        updatedAt: '2024-05-17T22:15:58.521Z',
      }),
    );
    expect(data?.[1]).toEqual(
      expect.objectContaining({
        phoneNumber: '333-3333',
        userId: 'some-other-user-id',
        createdAt: '2024-05-17T22:16:17.712Z',
        updatedAt: '2024-05-17T22:16:17.712Z',
      }),
    );
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('solo async handler', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          soloAsync: {
            success: true,
          },
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);

    const client = generateClient<Schema>();

    const { data } = await client.queries.soloAsync({
      value: 'hello, world!',
    });

    expect(data).toEqual(
      expect.objectContaining({
        success: true,
      }),
    );
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('async sync', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          asyncSync: {
            result: 'custom type echo result',
          },
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);

    const client = generateClient<Schema>();

    const { data } = await client.queries.asyncSync({
      value: 'hello, world!',
    });

    expect(data).toEqual({
      result: 'custom type echo result',
    });
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('sync sync', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          syncSync: {
            result: 'custom type echo result',
          },
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);

    const client = generateClient<Schema>();

    const { data } = await client.queries.syncSync({
      value: 'hello, world!',
    });

    expect(data).toEqual({
      result: 'custom type echo result',
    });
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('sync async', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          syncAsync: {
            success: true,
          },
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);

    const client = generateClient<Schema>();

    const { data } = await client.queries.syncAsync({
      value: 'hello, world!',
    });

    expect(data).toEqual(
      expect.objectContaining({
        success: true,
      }),
    );
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  test('async async', async () => {
    const { spy, generateClient } = mockedGenerateClient([
      {
        data: {
          asyncAsync: {
            success: true,
          },
        },
      },
    ]);

    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);

    const client = generateClient<Schema>();

    const { data } = await client.queries.asyncAsync({
      value: 'hello, world!',
    });

    expect(data).toEqual(
      expect.objectContaining({
        success: true,
      }),
    );
    expect(optionsAndHeaders(spy)).toMatchSnapshot();
  });

  describe('with enum arguments', () => {
    const schema = a.schema({
      Status: a.enum(['Active', 'Inactive', 'Unknown']),
      echoEnum: a
        .query()
        .arguments({
          status: a.enum(['Active', 'Inactive', 'Unknown']),
        })
        .returns(a.ref('Status'))
        .handler(a.handler.function('name' as any))
        .authorization((allow) => [allow.publicApiKey()]),
    });

    type Schema = ClientSchema<typeof schema>;

    test(`appear with enums values in the ClientSchema['args']`, async () => {
      type ExpectedArgs = {
        status?: 'Active' | 'Inactive' | 'Unknown' | null | undefined;
      };
      type test = Expect<Equal<Schema['echoEnum']['args'], ExpectedArgs>>;
    });

    test('appear in the schema model', async () => {
      /**
       * Relevant invariants:
       *
       * 1. `echoEnum` must have arguments `(status: SOME_ENUM_TYPE_NAME)`
       * 2. `SOME_ENUM_TYPE_NAME` must appear as an enum containing `Active, Inactive, Unknown`
       */
      expectGraphqlMatches(
        schema.transform().schema,
        `
        enum Status {
          Active,
          Inactive,
          Unknown
        }

        enum EchoEnumStatus {
          Active,
          Inactive,
          Unknown
        }

        type Query {
          echoEnum(status: EchoEnumStatus): Status @function(name: "name") @auth(rules: [{allow: public, provider: apiKey}])
        }
      `,
      );
    });

    test('produce operation-namespaced enums in the modelIntrospection schema', async () => {
      const { modelIntrospection } = await buildAmplifyConfig(schema);
      expect(modelIntrospection.enums.EchoEnumStatus).toEqual(
        expect.objectContaining({
          name: 'EchoEnumStatus',
          values: ['Active', 'Inactive', 'Unknown'],
        }),
      );
    });

    test('produce query entries referring to the namespaced enum in the modelIntrospection schema', async () => {
      const { modelIntrospection } = await buildAmplifyConfig(schema);
      expect(modelIntrospection.queries.echoEnum.arguments).toEqual(
        expect.objectContaining({
          status: {
            name: 'status',
            isArray: false,
            type: { enum: 'EchoEnumStatus' },
            isRequired: false,
          },
        }),
      );
    });

    test('can be called with valid enum values', async () => {
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: { echoEnum: 'Active' },
        },
      ]);

      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();

      const { data } = await client.queries.echoEnum({ status: 'Active' });
      expect(optionsAndHeaders(spy)).toMatchSnapshot();
      expect(data).toEqual('Active');
    });

    test('raise type errors when called with invalid enum values', async () => {
      const { spy, generateClient } = mockedGenerateClient([
        {
          data: { echoEnum: 'Active' },
        },
      ]);

      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();

      // @ts-expect-error
      const { data } = await client.queries.echoEnum({ status: 'BAD VALUE' });
    });
  });
});
