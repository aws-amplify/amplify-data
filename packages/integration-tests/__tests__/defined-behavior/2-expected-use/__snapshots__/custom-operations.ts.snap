// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom operations custom type array result 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    query($value: String) {
      multiEcho(value: $value) {result}
    }
  ",
      "variables": {
        "value": "something",
      },
    },
    {},
  ],
]
`;

exports[`custom operations custom type result 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    query($value: String) {
      echoWithCustomType(value: $value) {result}
    }
  ",
      "variables": {
        "value": "something",
      },
    },
    {},
  ],
]
`;

exports[`custom operations model array result 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    query($phoneNumbers: [String]) {
      checkBatchOfPhoneNumbersForActiveUsers(phoneNumbers: $phoneNumbers) {phoneNumber createdAt updatedAt}
    }
  ",
      "variables": {
        "phoneNumbers": [
          "123",
          "123",
        ],
      },
    },
    {},
  ],
]
`;

exports[`custom operations model result 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    query($query: String) {
      findJustOnePhoneNumber(query: $query) {phoneNumber createdAt updatedAt}
    }
  ",
      "variables": {
        "query": "something",
      },
    },
    {},
  ],
]
`;

exports[`custom operations primitive type result 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    query($value: String) {
      echo(value: $value) 
    }
  ",
      "variables": {
        "value": "something",
      },
    },
    {},
  ],
]
`;

exports[`custom operations with enum arguments can be called with valid enum values 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "
    query($status: EchoEnumStatus) {
      echoEnum(status: $status) 
    }
  ",
      "variables": {
        "status": "Active",
      },
    },
    {},
  ],
]
`;
